### ISSUE TITLE
Is there a mechanism to filter exception serialization?

### ISSUE THREAD

**Issue by mmurrell** [NLog/NLog#3588]
### Summary
We recently encountered a scenario in NLog where we logged what happened to be an Entity Framework exception, and we did so  with full serialization.  The Entity Framework exception included references to the database entity in question,  and while serializing the object graph of the exception NLog began to evaluate every property which caused a 'Lazy Load' on all the related tables, their related tables, and their related tables.  This particular event was responsible for an application outage.

The particular exception that we attempted to log was a `UpdateException` ([link](https://docs.microsoft.com/en-us/dotnet/api/system.data.entity.core.updateexception?view=entity-framework-6.2.0)).  The serialization problem begins when serializing `exception.StateEntries[0].Entity`.  

As a consumer, it would be nice to blacklist certain types using a lambda, or another mechanism, or give the consumer the ability to subclass the serializer.  The classes would be trivial to exclude by lambda, because the class names all follow the pattern `System.Data.Entity.DynamicProxies.TableName_E2DC4A5B7B1B73EDD543E8CD945EC967B53FA31EA65D6F4D4F23EC13451D87FD`

### Details
**Version:** NLog 4.6.3 via Nuget
**Configuration:**
```
<nlog>
  <targets>
    <target name="console" type="Console" layout="${sequenceid}|${machinename}|${longdate}|${logger}|${level}|${message}|${exception:format=@}"/>
  </targets>
  <rules>
    <logger name="*" minlevel="Info" writeTo="console"/>
  </rules>
</nlog>
```

**Sample Code:**
```
try
{
    // Database configured to have multiple tables, lazy loading enabled, TableName referencing many child tables.
    var db = IoC.Container.GetInstance<EntityFrameworkDataContext>()
    var databaseObject = db.TableName.First();
    databaseObject.NonNullableField = null;
    db.SubmitChanges();
}
catch (Exception ex)
{
    // On execution, run Sql Server Profiler and view all tables in the database being scanned.
    Logger.Error(ex, "Error");
}
```
**Internal Log (Debug):** Available if you need it, but it doesn't show anything relevant
**Current Result:** When `Logger.Error(exception, "...")` is called with a Entity Framework exception, the entire object graph is traversed by the NLog Serializer, causing repeated table scans against the entire database.
**Expected Result:** A mechanism to blacklist certain types from being serialized.
**Workarounds:** 
- Remove exception from the error log: `Logger.Error("Something happened." + ex.Message);`,
- Modify every place that we could potentially see an EF Exception with a seperate set of logging. (This would involve replacing every exception handling scenario in the application with a second catch)
```
// Should replace this:
catch(Exception ex){ Logger.Error(ex, "..."); }
// with this:
catch(UpdateException uex){ Logger.Error("UpdateException..."); }
catch(Exception ex){ Logger.Error(ex, "..."); }
```

Any help you could offer would be greatly appreciated.

**Comment by snakefoot** [comment]
Also interested in making the Exception-serialization as safe as possible.

You can always overrider the JsonConverter in NLog, and perform special handling of `System.Data.DataException`-types:

Maybe this will work:
```c#
internal class EvilExceptionJsonConverter : NLog.IJsonConverter
{
    private JsonSerializeOptions _jsonOptions = new NLog.Targets.JsonSerializeOptions() { MaxRecursionLimit = 1 };
    private NLog.IJsonConverter _originalConverter;
    public JsonLogModelSerializer(NLog.IJsonConverter originalConverter)
    {
        _originalConverter = originalConverter;
    }

    /// <summary>Serialization of an object into JSON format.</summary>
    /// <param name="value">The object to serialize to JSON.</param>
    /// <param name="builder">Output destination.</param>
    /// <returns>Serialize succeeded (true/false)</returns>
    public bool SerializeObject(object value, System.Text.StringBuilder builder)
    {
        if (value is System.Data.DataException evilException)
        {
            return NLog.Targets.DefaultJsonSerializer.Instance.SerializeObject(value, builder, _jsonOptions);
        }

        return _originalConverter.SerializeObject(value, builder);
    }
}
```

Then register it like this, before loading NLog.config:
```C#
var oldJsonConverter = NLog.Config.ConfigurationItemFactory.Default.JsonConverter;
var newConverter = new EvilExceptionJsonConverter(oldJsonConverter);
NLog.Config.ConfigurationItemFactory.Default.JsonConverter = newConverter;
```

**Comment by snakefoot** [comment]
@mmurrell Would it work if changed the default JsonConverter to handle Exception-properties of type IEnumerable a little different?

Instead of doing reflection on objects inside IEnumerable, then it will just perform ToString() of each item. What happens if one calls `ObjectStateEntry.ToString()` ?

**Comment by mmurrell** [comment]
@snakefoot - Thanks for the quick reply.

I'd hate to limit the serializer to "ToString" enumerable properties because I can imagine several "safe" scenarios where the enumerable data inside an exception, would add value... probably most notably `AggregateException`. 

Your sample `EvilExceptionJsonConverter` may be a good short-term workaround to prevent this specific crash.  The underlying concern is that when there serializer has the max depth set, the logs would be missing the detail that could be serialized safely,  

Would it be possible to "unseal" (make the singleton pattern publicly settable, and some of the methods protected) so a consumer could hook into the recursive SerializeObject calls? or perhaps if this were worth supporting at a NLog level, would you consider a `Func<object, bool> ShouldSerializeObject` on the JsonSerializeOptions that can be consumer-set?

**Comment by snakefoot** [comment]
I guess we could also make it possible to pre-register known dangerous types in the `ObjectReflectionCache`.

The NLog `ObjectReflectionCache` controls what properties to include in reflection. I have already implemented rules to not reflect dangerous types like these:

```
            if (typeof(IFormattable).IsAssignableFrom(objectType))
                return true;

            if (typeof(Uri).IsAssignableFrom(objectType))
                return true;

            if (typeof(MemberInfo).IsAssignableFrom(objectType))
                return true;

            if (typeof(Assembly).IsAssignableFrom(objectType))
                return true;
```

I guess one could create an interface to make it possible to register a type like `ObjectStateEntry`

**Comment by 304NotModified** [comment]
> I guess we could also make it possible to pre-register known dangerous types in the `ObjectReflectionCache`.

Sounds good to me!

**Comment by snakefoot** [comment]
Resolved with  #3610

**Comment by snakefoot** [comment]
@mmurrell NLog 4.7 has been released that includes #3610
